core/src/main/java/hudson/FilePath.java

```diff
@@ -478,7 +478,18 @@ public int zip(OutputStream out, DirScanner scanner) throws IOException, Interru
      */
     public int archive(final ArchiverFactory factory, OutputStream os, final DirScanner scanner) throws IOException, InterruptedException {
         final OutputStream out = (channel!=null)?new RemoteOutputStream(os):os;
-        return act(new SecureFileCallable<Integer>() {
+        return act(new Archive(factory, out, scanner));
+    }
+    private class Archive extends SecureFileCallable<Integer> {
+        private final ArchiverFactory factory;
+        private final OutputStream out;
+        private final DirScanner scanner;
```

Comment:
It's a bad idea to serialize it over the channel IMHO. Add better APIs and deprecate the existing method?

Reply:
The `DirScanner` needs to be serialized by the nature of this API.